File: silmari_rlm_act/phases/implementation.py
"""Implementation phase - Simple Claude loop.

This module implements the implementation phase using a simple loop pattern:
1. Inject beads epic/issues into the TDD plan
2. Pass the plan to Claude with --dangerously-skip-permissions
3. Let Claude manage implementation, tests, commits, and beads
4. Check for completion after each iteration
5. Stop when all beads issues are closed

Claude is powerful - let Claude manage the implementation.
"""

import subprocess
import time
from datetime import datetime
from pathlib import Path
from typing import Any, Optional

from planning_pipeline.claude_runner import run_claude_subprocess
from silmari_rlm_act.context.cwa_integration import CWAIntegration
from silmari_rlm_act.models import AutonomyMode, PhaseResult, PhaseStatus, PhaseType


class ImplementationPhase:
    """Execute TDD implementation via simple Claude loop.

    Pattern from loop.sh:
        while true; do
            cat PLAN.md | claude -p --dangerously-skip-permissions
            sleep 10
            # check for completion
        done

    Attributes:
        project_path: Root directory of the project
        cwa: Context Window Array integration
        LOOP_SLEEP: Seconds between loop iterations
        CLAUDE_TIMEOUT: Max seconds per Claude invocation
        TEST_TIMEOUT: Max seconds for test suite execution
    """

    LOOP_SLEEP = 10
    CLAUDE_TIMEOUT = 3600  # 1 hour per iteration
    TEST_TIMEOUT = 900  # 15 minutes for full test suite

    def __init__(
        self,
        project_path: Path,
        cwa: CWAIntegration,
    ) -> None:
        """Initialize implementation phase.

        Args:
            project_path: Root directory of the project
            cwa: Context Window Array integration instance
        """
        self.project_path = Path(project_path)
        self.cwa = cwa

    def _build_implementation_prompt(
        self,
        plan_path: Path,
        epic_id: Optional[str],
        issue_ids: list[str],
    ) -> str:
        """Build prompt that tells Claude to read and implement the TDD plan.

        Args:
            plan_path: Path to the TDD plan overview file
            epic_id: Optional beads epic ID
            issue_ids: List of beads issue IDs for phases

        Returns:
            Prompt instructing Claude to implement the plan
        """
        prompt = f"""Implement the TDD plan at: {plan_path}

Read the plan overview first, then find and implement the phase documents.

## Beads Tracking

Use `bd` commands to track progress:
```bash
bd ready                    # See available work
bd show <id>                # View issue details
bd update <id> --status=in_progress  # Start work
bd close <id>               # Complete work (unblocks next)
bd sync                     # Sync changes
```

"""
        if epic_id:
            prompt += f"**Epic**: `{epic_id}`\n\n"

        if issue_ids:
            prompt += "**Phase Issues**:\n"
            for i, issue_id in enumerate(issue_ids):
                prompt += f"- Phase {i + 1}: `{issue_id}`\n"
            prompt += "\n"

        prompt += """
## Implementation Instructions

1. Read the plan overview at the path above
2. Find the phase documents in the same directory
3. Implement the highest priority TASK using subagents
4. Run all tests: `pytest` or `make test`
5. Update the plan with progress
6. Use `bd close <id>` when phase is complete
7. Use `/clear` after closing an issue to start fresh

**CRITICAL**: After ALL TESTS PASS and after each successful `bd close`,
emit a /clear command to clear context for the next issue.
"""
        return prompt

    def _invoke_claude(self, prompt: str) -> dict[str, Any]:
        """Invoke Claude with plan as prompt.

        Uses run_claude_subprocess for real-time streaming output to terminal.

        Args:
            prompt: The full prompt (plan content with beads)

        Returns:
            Dict with success, output, error, elapsed
        """
        return run_claude_subprocess(
            prompt,
            timeout=self.CLAUDE_TIMEOUT,
            stream_json=False,  # Text mode for human-readable output
            cwd=str(self.project_path),
        )

    def _check_completion(self, issue_ids: list[str]) -> bool:
        """Check if all beads issues are closed.

        Args:
            issue_ids: List of issue IDs to check

        Returns:
            True if all issues are closed
        """
        if not issue_ids:
            return True

        try:
            # Check each issue status
            for issue_id in issue_ids:
                result = subprocess.run(
                    ["bd", "show", issue_id],
                    cwd=str(self.project_path),
                    capture_output=True,
                    text=True,
                    timeout=30,
                )

                output = result.stdout.lower()
                # If any issue is not closed, return False
                if "status: closed" not in output and "status: done" not in output:
                    return False

            return True

        except Exception:
            # If check fails, assume not complete
            return False

    def _run_tests(self) -> tuple[bool, str]:
        """Run test suite to verify implementation.

        Returns:
            Tuple of (passed, output)
        """
        try:
            result = subprocess.run(
                ["pytest", "-v", "--tb=short"],
                cwd=str(self.project_path),
                capture_output=True,
                text=True,
                timeout=self.TEST_TIMEOUT,
            )
            return result.returncode == 0, result.stdout + result.stderr

        except subprocess.TimeoutExpired:
            return False, f"Tests timed out after {self.TEST_TIMEOUT} seconds"
        except FileNotFoundError:
            # Try make test as fallback
            try:
                result = subprocess.run(
                    ["make", "test"],
                    cwd=str(self.project_path),
                    capture_output=True,
                    text=True,
                    timeout=self.TEST_TIMEOUT,
                )
                return result.returncode == 0, result.stdout + result.stderr
            except Exception:
                return True, "No test command found, skipping"
        except Exception as e:
            return False, str(e)

    def execute(
        self,
        phase_paths: list[str],
        mode: AutonomyMode,  # noqa: ARG002 - kept for API compatibility
        beads_issue_ids: Optional[list[str]] = None,
        beads_epic_id: Optional[str] = None,
        max_iterations: int = 100,
        checkpoint: Optional[dict[str, Any]] = None,  # noqa: ARG002 - kept for API compatibility
    ) -> PhaseResult:
        """Execute implementation via simple Claude loop.

        Args:
            phase_paths: Paths to TDD plan documents
            mode: Execution mode (ignored - always autonomous)
            beads_issue_ids: Beads issue IDs for tracking
            beads_epic_id: Optional beads epic ID
            max_iterations: Safety limit on loop iterations
            checkpoint: Optional checkpoint (ignored)

        Returns:
            PhaseResult with implementation status
        """
        # Suppress unused parameter warnings - kept for API compatibility
        _ = mode, checkpoint

        started_at = datetime.now()
        errors: list[str] = []
        artifacts: list[str] = []
        issue_ids = beads_issue_ids or []

        if not phase_paths:
            return PhaseResult(
                phase_type=PhaseType.IMPLEMENTATION,
                status=PhaseStatus.COMPLETE,
                artifacts=[],
                started_at=started_at,
                completed_at=datetime.now(),
                duration_seconds=0,
                metadata={"message": "No phase paths provided"},
            )

        # Use first plan as the main plan (overview)
        plan_path = Path(phase_paths[0])
        if not plan_path.is_absolute():
            plan_path = self.project_path / plan_path

        if not plan_path.exists():
            return PhaseResult(
                phase_type=PhaseType.IMPLEMENTATION,
                status=PhaseStatus.FAILED,
                errors=[f"Plan not found: {plan_path}"],
                started_at=started_at,
                completed_at=datetime.now(),
                duration_seconds=0,
            )

        # Build prompt with plan path - Claude will read the plan
        prompt = self._build_implementation_prompt(plan_path, beads_epic_id, issue_ids)

        # Simple loop: invoke Claude, sleep, check completion
        iteration = 0
        while iteration < max_iterations:
            iteration += 1
            print(f"\n{'=' * 60}")
            print(f"IMPLEMENTATION LOOP - Iteration {iteration}")
            print(f"{'=' * 60}\n")

            # Invoke Claude with plan (output streams directly to terminal)
            result = self._invoke_claude(prompt)

            if not result["success"]:
                print(f"Claude invocation failed: {result.get('error', 'unknown error')}")

            # Sleep between iterations
            print(f"\n{'=' * 25} LOOP {'=' * 25}\n")
            time.sleep(self.LOOP_SLEEP)

            # Check if all issues are closed
            if self._check_completion(issue_ids):
                print("All beads issues closed - implementation complete!")

                # Final test verification
                print("Running final test verification...")
                tests_pass, test_output = self._run_tests()
                if not tests_pass:
                    print(f"Tests failed - continuing loop. Error: {test_output[:200]}")
                    errors.append(f"Final tests failed: {test_output[:500]}")
                    # Continue loop if tests fail
                    continue

                # Tests passed - clear any previous test failures and exit successfully
                print("Tests passed! Implementation complete.")
                errors.clear()
                artifacts.extend(phase_paths)
                break

        else:
            errors.append(f"Reached max iterations ({max_iterations})")

        # Loop exited - print summary
        if not errors:
            print(f"\n{'=' * 60}")
            print("SUCCESS: Implementation loop completed successfully!")
            print(f"Total iterations: {iteration}")
            print(f"{'=' * 60}\n")
        else:
            print(f"\n{'=' * 60}")
            print("FAILED: Implementation loop did not complete successfully")
            print(f"Errors: {errors}")
            print(f"{'=' * 60}\n")

        completed_at = datetime.now()
        duration = (completed_at - started_at).total_seconds()

        return PhaseResult(
            phase_type=PhaseType.IMPLEMENTATION,
            status=PhaseStatus.COMPLETE if not errors else PhaseStatus.FAILED,
            artifacts=artifacts,
            errors=errors,
            started_at=started_at,
            completed_at=completed_at,
            duration_seconds=duration,
            metadata={
                "iterations": iteration,
                "mode": "autonomous_loop",
                "beads_issues": issue_ids,
            },
        )

File: go/internal/planning/implementation.go
package planning

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

// Constants for the implementation loop
const (
	IMPL_LOOP_SLEEP     = 10 * time.Second
	IMPL_MAX_ITERATIONS = 100
	IMPL_TIMEOUT        = 3600 // 1 hour per iteration in seconds
	TEST_TIMEOUT        = 300  // 5 minutes for test execution
)

// ImplementationResult contains the result of the implementation phase.
type ImplementationResult struct {
	Success       bool     `json:"success"`
	Error         string   `json:"error,omitempty"`
	Iterations    int      `json:"iterations"`
	TestsPassed   bool     `json:"tests_passed"`
	PhasesClosed  []string `json:"phases_closed,omitempty"`
	Output        string   `json:"output,omitempty"`
}

// StepImplementation executes the autonomous implementation phase.
// This function implements the core TDD autonomous loop that:
// 1. Invokes Claude with the implementation prompt
// 2. Sleeps between iterations
// 3. Checks if all beads issues are closed
// 4. Runs tests when issues are closed
// 5. Returns success when tests pass
func StepImplementation(
	projectPath string,
	phasePaths []string,
	beadsIssueIDs []string,
	beadsEpicID string,
	maxIterations int,
) *ImplementationResult {
	result := &ImplementationResult{Success: true}

	// Set max iterations to default if not provided
	if maxIterations == 0 {
		maxIterations = IMPL_MAX_ITERATIONS
	}

	// Validate inputs
	if projectPath == "" {
		result.Success = false
		result.Error = "project_path is required"
		return result
	}

	if len(beadsIssueIDs) == 0 {
		result.Success = false
		result.Error = "beads_issue_ids is required"
		return result
	}

	// Build implementation prompt
	prompt := buildImplementationPrompt(phasePaths, beadsEpicID, beadsIssueIDs)

	// Start implementation loop
	fmt.Printf("\n=== Starting Implementation Loop (max %d iterations) ===\n", maxIterations)

	for i := 0; i < maxIterations; i++ {
		result.Iterations = i + 1

		fmt.Printf("\n--- Iteration %d/%d ---\n", result.Iterations, maxIterations)

		// Invoke Claude with the implementation prompt
		claudeResult := RunClaudeSync(prompt, IMPL_TIMEOUT, true, projectPath)
		if !claudeResult.Success {
			// Log error but continue loop (may be transient)
			fmt.Printf("WARNING: Claude iteration %d failed: %s\n", result.Iterations, claudeResult.Error)
			fmt.Println("Continuing to next iteration...")
		}

		// Capture output
		if claudeResult.Output != "" {
			result.Output += fmt.Sprintf("\n=== Iteration %d ===\n%s", result.Iterations, claudeResult.Output)
		}

		// Sleep before checking issue status
		fmt.Printf("Sleeping for %v before checking status...\n", IMPL_LOOP_SLEEP)
		time.Sleep(IMPL_LOOP_SLEEP)

		// Check if all beads issues are closed
		fmt.Println("Checking if all beads issues are closed...")
		allClosed, closedIssues := checkAllIssuesClosed(projectPath, beadsIssueIDs)
		result.PhasesClosed = closedIssues

		if allClosed {
			fmt.Println("All beads issues are closed! Running tests...")

			// Run tests
			testsPassed, testOutput := runTests(projectPath)
			result.TestsPassed = testsPassed

			if testsPassed {
				fmt.Println("SUCCESS: All tests passed!")
				result.Success = true
				result.Output += fmt.Sprintf("\n=== Test Output ===\n%s", testOutput)
				return result
			}

			fmt.Printf("Tests failed. Continuing loop to fix issues.\n")
			fmt.Printf("Test output:\n%s\n", testOutput)

			// Update prompt to include test failures
			prompt = buildImplementationPrompt(phasePaths, beadsEpicID, beadsIssueIDs) +
				fmt.Sprintf("\n\n## Test Failures from Previous Iteration\n```\n%s\n```\n\nFix these test failures and ensure all tests pass.\n", testOutput)
		} else {
			fmt.Printf("Not all issues closed yet. Open issues: %v\n", getOpenIssues(beadsIssueIDs, closedIssues))
		}
	}

	// Max iterations reached without success
	result.Success = false
	result.Error = fmt.Sprintf("max iterations (%d) reached without completing implementation", maxIterations)
	return result
}

// buildImplementationPrompt constructs the prompt for Claude to implement the TDD phases.
func buildImplementationPrompt(phasePaths []string, epicID string, issueIDs []string) string {
	var sb strings.Builder

	sb.WriteString("# TDD Implementation Task\n\n")
	sb.WriteString("You are implementing a TDD plan. Follow the Red-Green-Refactor cycle for each behavior.\n\n")

	// Add phase file references
	if len(phasePaths) > 0 {
		sb.WriteString("## TDD Plan Files\n")
		for i, path := range phasePaths {
			sb.WriteString(fmt.Sprintf("%d. Read the plan at: %s\n", i+1, path))
		}
		sb.WriteString("\n")
	}

	// Add beads epic reference
	if epicID != "" {
		sb.WriteString("## Beads Epic\n")
		sb.WriteString(fmt.Sprintf("Epic ID: %s\n", epicID))
		sb.WriteString(fmt.Sprintf("View: `bd show %s`\n\n", epicID))
	}

	// Add phase issues
	if len(issueIDs) > 0 {
		sb.WriteString("## Phase Issues\n")
		sb.WriteString("Complete these phases in order:\n")
		for i, id := range issueIDs {
			sb.WriteString(fmt.Sprintf("- Phase %d: %s (`bd show %s`)\n", i+1, id, id))
		}
		sb.WriteString("\n")
	}

	// Add implementation instructions
	sb.WriteString(`## Implementation Instructions

Follow these steps for EACH phase:

1. **Read the Phase Plan**: Carefully read the phase file to understand:
   - Requirements and testable behaviors
   - Success criteria
   - Dependencies

2. **Red-Green-Refactor Cycle**: For each testable behavior:
   a. Write a failing test (RED)
   b. Implement minimal code to make it pass (GREEN)
   c. Refactor if needed while keeping tests green (REFACTOR)

3. **Run Tests**: After implementing behaviors, run the test suite:
   - Use: pytest -v --tb=short
   - Or: make test
   - Ensure all tests pass before proceeding

4. **Close Phase Issue**: When a phase is complete and tests pass:
   - Close the issue: bd close <issue-id>
   - Add a comment with summary: bd comment <issue-id> "Implemented X behaviors, all tests pass"

5. **Clear Context**: After closing an issue, emit /clear to reset context

6. **Move to Next Phase**: Continue with the next phase issue

## Critical Rules

- ALWAYS run tests before closing an issue
- ALWAYS close issues when phase is complete
- ALWAYS emit /clear after closing issues
- DO NOT skip tests
- DO NOT close issues unless ALL behaviors are implemented and tests pass
- If tests fail, fix them before closing the issue

## Exit Conditions

The implementation is complete when:
1. All phase issues are closed
2. All tests pass
3. No open beads issues remain

Continue working until these conditions are met.
`)

	return sb.String()
}

// checkAllIssuesClosed checks if all beads issues are closed.
// Returns (allClosed bool, closedIssueIDs []string).
func checkAllIssuesClosed(projectPath string, issueIDs []string) (bool, []string) {
	if len(issueIDs) == 0 {
		return false, []string{}
	}

	closedIssues := []string{}

	for _, id := range issueIDs {
		if isIssueClosed(projectPath, id) {
			closedIssues = append(closedIssues, id)
		}
	}

	return len(closedIssues) == len(issueIDs), closedIssues
}

// isIssueClosed checks if a single beads issue is closed.
func isIssueClosed(projectPath, issueID string) bool {
	cmd := exec.Command("bd", "show", issueID)
	cmd.Dir = projectPath
	output, err := cmd.Output()
	if err != nil {
		// If bd command fails, assume issue is not closed
		return false
	}

	outputStr := strings.ToLower(string(output))

	// Check for common "closed" status indicators
	closedIndicators := []string{
		"status: closed",
		"status: done",
		"status: complete",
		"status:closed",
		"status:done",
		"status:complete",
	}

	for _, indicator := range closedIndicators {
		if strings.Contains(outputStr, indicator) {
			return true
		}
	}

	return false
}

// getOpenIssues returns the list of open issue IDs.
func getOpenIssues(allIssues, closedIssues []string) []string {
	closedSet := make(map[string]bool)
	for _, id := range closedIssues {
		closedSet[id] = true
	}

	openIssues := []string{}
	for _, id := range allIssues {
		if !closedSet[id] {
			openIssues = append(openIssues, id)
		}
	}

	return openIssues
}

// runTests executes the test suite and returns (passed bool, output string).
// Implements REQ_012.3: pytest-first strategy with fallback to make test only when pytest binary not found.
func runTests(projectPath string) (bool, string) {
	fmt.Println("Attempting to run tests with pytest...")

	// Try pytest first (REQ_012.1)
	passed, output, err := tryPytest(projectPath)

	// If pytest ran (even if tests failed), return the result
	// REQ_012.2: Only fallback when pytest binary is not found
	if err == nil {
		fmt.Printf("Tests executed with pytest. Passed: %v\n", passed)
		return passed, output
	}

	// Check if error is due to pytest not being found
	if err == exec.ErrNotFound || strings.Contains(err.Error(), "executable file not found") {
		fmt.Println("pytest not found, falling back to make test...")

		// Fallback to make test (REQ_012.2)
		passed, output = tryMakeTest(projectPath)
		if output != "" {
			fmt.Printf("Tests executed with make test. Passed: %v\n", passed)
			return passed, output
		}

		// Neither pytest nor make test available
		fmt.Println("No test command found (pytest and make test not available), skipping tests")
		return true, "No test command found, skipping"
	}

	// pytest exists but execution failed for another reason
	fmt.Printf("pytest execution failed: %v\n", err)
	return false, fmt.Sprintf("pytest execution error: %v\n%s", err, output)
}

// tryPytest attempts to run pytest with timeout.
// Implements REQ_012.1: Execute pytest as primary test command with verbose output.
// Returns (passed bool, output string, error).
// Error is non-nil only when pytest binary is not found or execution fails.
func tryPytest(projectPath string) (bool, string, error) {
	// Check if pytest is available first
	checkCmd := exec.Command("pytest", "--version")
	if err := checkCmd.Run(); err != nil {
		// pytest binary not found - return error to trigger fallback
		return false, "", err
	}

	// Run pytest with verbose output and short traceback (REQ_012.1)
	cmd := exec.Command("pytest", "-v", "--tb=short")
	cmd.Dir = projectPath

	// Create a channel to capture command result
	type cmdResult struct {
		output []byte
		err    error
	}
	resultChan := make(chan cmdResult, 1)

	// Run command in goroutine with timeout (REQ_012.1)
	go func() {
		output, err := cmd.CombinedOutput()
		resultChan <- cmdResult{output: output, err: err}
	}()

	// Wait for command to complete or timeout
	select {
	case result := <-resultChan:
		// Command completed within timeout
		// REQ_012.1: Return exit code 0 as success, non-zero as failure
		passed := result.err == nil
		return passed, string(result.output), nil

	case <-time.After(TEST_TIMEOUT * time.Second):
		// Timeout occurred (REQ_012.1)
		// Try to kill the process if it's still running
		if cmd.Process != nil {
			_ = cmd.Process.Kill()
		}
		return false, "", fmt.Errorf("pytest execution timed out after %d seconds", TEST_TIMEOUT)
	}
}

// tryMakeTest attempts to run make test with timeout.
// Implements REQ_012.2: Execute 'make test' as fallback when pytest not available.
func tryMakeTest(projectPath string) (bool, string) {
	// Check if Makefile exists
	makefilePath := filepath.Join(projectPath, "Makefile")
	if _, err := os.Stat(makefilePath); os.IsNotExist(err) {
		return false, ""
	}

	// Run make test with same timeout as pytest (REQ_012.2)
	cmd := exec.Command("make", "test")
	cmd.Dir = projectPath

	// Create a channel to capture command result
	type cmdResult struct {
		output []byte
		err    error
	}
	resultChan := make(chan cmdResult, 1)

	// Run command in goroutine with timeout
	go func() {
		output, err := cmd.CombinedOutput()
		resultChan <- cmdResult{output: output, err: err}
	}()

	// Wait for command to complete or timeout
	select {
	case result := <-resultChan:
		// Command completed within timeout
		passed := result.err == nil
		return passed, string(result.output)

	case <-time.After(TEST_TIMEOUT * time.Second):
		// Timeout occurred
		if cmd.Process != nil {
			_ = cmd.Process.Kill()
		}
		return false, fmt.Sprintf("make test execution timed out after %d seconds", TEST_TIMEOUT)
	}
}

